libraries/chain/include/eosio/chain/contract_types.hpp

```diff
@@ -211,6 +229,23 @@ struct onerror {
    }
 };
 
+struct calcbw {
```

Comment:
reading this name, the question arises why consider the bandwidth.
Maybe better name this structure and action `approvebw`?
Then we get a coherent system: the user requests a bandwidth with `requestbw` action, then provider's contract approve it with `approvebw` (and indicates how much bandwidth can be spent).